# Part III: Ensembles and Final Result

## AdaBoost

Train an AdaBoost classifier using Decision Tree stubs as weak learners. Compare its performance to results obtained in Part II using 10 fold CV.

In [1]:
import proj2_lib.util as utils

In [2]:
utils.file_config

{'feature_pipeline_file': 'feature_pipeline.pkl',
 'labels_pipeline_file': 'labels_pipeline.pkl',
 'objstore_path': 'objects',
 'processed_data_path': 'processed_data',
 'raw_data_csv': 'KaggleV2-May-2016.csv',
 'raw_data_path': 'data',
 'test_csv': 'test_set.csv',
 'train_csv': 'train_set.csv'}

In [3]:
file_config = utils.file_config

In [4]:
# ONLY NEED TO RUN THIS STEP ONCE (switch this to True to run it)
RUN_MAKE_TRAIN_TEST_FILES = False
if RUN_MAKE_TRAIN_TEST_FILES:
    utils.make_train_test_sets(config=file_config)

In [5]:
import proj2_lib.preprocess as preprocess

# ONLY NEED TO RUN THIS STEP ONCE
RUN_FIT_PREPROCESSING = False
if RUN_FIT_PREPROCESSING:
    preprocess.fit_save_pipelines(config=file_config)

In [6]:
train_X, train_y = preprocess.load_train_data(config=file_config)

In [7]:
print(train_X.shape)
print(train_y.shape)

(90526, 101)
(90526,)


In [8]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier

bdt = AdaBoostClassifier(DecisionTreeClassifier(max_depth=1),
                         algorithm="SAMME",
                         n_estimators=200)
bdt.fit(train_X,train_y)

AdaBoostClassifier(algorithm='SAMME',
          base_estimator=DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=1,
            max_features=None, max_leaf_nodes=None,
            min_impurity_split=1e-07, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            presort=False, random_state=None, splitter='best'),
          learning_rate=1.0, n_estimators=200, random_state=None)

In [10]:
# Calculate accuracy score using 10-fold cross-validation
from sklearn.model_selection import cross_val_score
scores = cross_val_score(bdt, train_X, train_y, 
                         scoring='accuracy', cv=10)

In [11]:
# mean accuracy of AdaBoosted Decision Tree classifier

import numpy as np
np.mean(scores)

0.79791440170735117

The accuracy of the AdaBoosted Decision Tree classifier (~79.8%) > the accuracy of Linear SVM (~79.6%) > the accuracy of the Random Forest Classifier (~77.7%) > the accuracy of Decision Tree classifier (~73.8%).

In [13]:
auc_scores = cross_val_score(bdt, train_X, train_y,
                            scoring='roc_auc', cv=10)

In [14]:
# mean AUC score of AdaBoosted Decision Tree classifier
np.mean(auc_scores)

0.7257816926656615

The AUC score of the AdaBoosted Decision Tree classifier (~72.6%) > the AUC score of the Random Forest classifier (~69.3%) > the AUC score of the Linear SVM (~66.7%) > the AUC score of the Decision Tree classifier (~58.3%).

## Stacking

Choose a set of 5 or so classifiers. Write a function that trains an ensemble using stacking

In [ ]:
def build_stack_ensemble(X, y):
    # create train/validation sets
    # using StratifiedShuffleSplit
    
    # train classifiers in ensemble using train set
    
    # create new feature matrix for validation
    # set by getting predictions from the ensemble
    # classifiers
    
    # train logistic regression classifier on
    # new feature matrix
    
    # return all trained classifiers
    return None

Use 10-fold cross validation to measure performance of your stacked classifier. See Part II solution to see how to roll your own sklearn classifier along with http://scikit-learn.org/stable/developers/contributing.html#rolling-your-own-estimator

## Final Result

Choose a single model based on all previous project steps. Train this model on the complete training dataset and measure it's performance on the held out test set.

Compare to the 10-fold CV estimate you got previously.

In [ ]:
# final result goes here